src/mavsdk_server/src/connection_initiator.h

```diff
@@ -45,29 +51,29 @@ template<typename Mavsdk> class ConnectionInitiator {
         return true;
     }
 
-    std::future<void> wrapped_subscribe_on_new_system(Mavsdk& mavsdk)
+    std::future<bool> wrapped_subscribe_on_new_system(Mavsdk& mavsdk)
     {
         auto future = _discovery_promise->get_future();
 
         mavsdk.subscribe_on_new_system([this, &mavsdk]() {
+            std::lock_guard<std::mutex> guard(_mutex);
             const auto system = mavsdk.systems().at(0);
 
-            if (system->is_connected()) {
-                std::call_once(_discovery_flag, [this]() {
-                    LogInfo() << "System discovered";
-                    _discovery_promise->set_value();
-                });
-            } else {
-                LogInfo() << "System timed out";
+            if (!_is_discovery_finished && system->is_connected()) {
+                LogInfo() << "System discovered";
+
+                _is_discovery_finished = true;
+                _discovery_promise->set_value(true);
             }
         });
 
         return future;
     }
 
-    std::once_flag _discovery_flag{};
-    std::shared_ptr<std::promise<void>> _discovery_promise{};
-    std::future<void> _discovery_future{};
+    std::mutex _mutex;
+    std::atomic<bool> _is_discovery_finished = false;
+    std::shared_ptr<std::promise<bool>> _discovery_promise{};
```

Comment:
Does this actually need to be a shared pointer? And why?

Reply:
Actually it's passed to a callback, so if the `ConnectionInitiator` object goes out of scope and the callback is called, we need a shared_ptr to avoid a crash. Or don't we?

Reply:
Ok, I guess so.